### PR TITLE
Removing some optimizations to prepare for refactoring

### PR DESCRIPTION
**Pull request by sebastienros**
- FunctionWasCalled is not used anymore but kept for future reference. It detects if a method is called to release an arguments array to the pull.
- The last binding memoization has been removed.

### CODE REVIEW

File: Jint/Runtime/Environments/Binding.cs [hunk 1]
```diff
@@ -14,5 +14,7 @@ public Binding(JsValue value, bool canBeDeleted, bool mutable)
         public JsValue Value;
         public readonly bool CanBeDeleted;
         public readonly bool Mutable;
+
+        public bool IsInitialized => !ReferenceEquals(Value, null);
```
Comment (lahma) [inline Jint/Runtime/Environments/Binding.cs]: Need to add a mental note about this non-readonly-struct and methods causes defensive copies, we can revisit when things settle.

**Comment by lahma** [comment]
Merge if ready for next steps, not sure if you are planning to do add onto this or smaller PRs.

**Comment by lahma** [comment]
UncacheableExpressionBenchmark shows taking twice the time and allocating 8x, so we probably need to get the tweaks (or their variants) back in when the dust settles.

**Comment by sebastienros** [comment]
Are you still ok with merging this one, I had forgotten about it, and the more we wait, the harder it will be. Unless your other PR would become hard to merge with these changes.

**Comment by lahma** [comment]
Sure thing, please merge, we can tweak performance later on. I'm more eager to get the let/const bits 🙂

**Comment by sebastienros** [comment]
Next step, before let/const work is done, is to get hoisting working based on the specs, and not based on the hackery that happens in esprima. And we can then remove it from esprima.

**Pull request merged**